opal/runtime/opal_params_core.c

```diff
@@ -350,3 +422,113 @@ int opal_register_util_params(void)
 
     return OPAL_SUCCESS;
 }
+
+/* Parses a color 'string' and extracts the 'code' for each 'key'
+ * The string is in the format of 'key1=code1:key2=code2:...'
+ * Example: opal_var_dump_color MCA parameter
+ *
+ * color_string: The string containing the key=code sequences to parse
+ * key_names: The keys to extract (e.g. key1, key2)
+ * key_count: The length of key_names
+ * values_out: The extracted values for the keys, 1-1 with key_names
+ *   ATTENTION: Make sure it's at least as large as key_names...
+ */
+static int parse_color_string(char *color_string, char **key_names,
+        int key_count, char **values_out) {
+
+    char **tokens = NULL, **kv = NULL;
+    int return_code = OPAL_SUCCESS;
+
+    for (int k = 0; k < key_count; k++) {
+        values_out[k] = NULL;
+    }
+
+    tokens = opal_argv_split(color_string, ',');
+    if (NULL == tokens) {
+        return_code = OPAL_ERR_OUT_OF_RESOURCE;
+        goto end;
+    }
+
+    for (int i = 0; tokens[i] != NULL; i++) {
+        kv = opal_argv_split(tokens[i], '=');
+        if (NULL == kv) {
+            return_code = OPAL_ERR_OUT_OF_RESOURCE;
+            goto end;
+        }
+
+        // Expected format of token: key=code
+        if (opal_argv_count(kv) != 2) {
+            opal_show_help("help-opal-runtime.txt", "mpi-params:var_dump_color:format-error",
+                true, tokens[i]);
+            goto skip_token;
+        }
+
+        bool key_found = false;
+
+        // Look for key name and store value in respective position
+        for (int k = 0; k < key_count; ++k) {
+            if (strcasecmp(key_names[k], kv[0]) == 0) {
+                int ret = opal_asprintf(&values_out[k], "\033[%sm", kv[1]);
+                if (ret < 0) {
+                    return_code = OPAL_ERR_OUT_OF_RESOURCE;
+                    goto end;
+                }
+
+                key_found = true;
+                break;
+            }
+        }
+
+        if (!key_found) {
+            char *valid_keys = NULL, *tmp;
+
+            for (int k = 0; k < key_count; k++) {
+                int ret = opal_asprintf(&tmp, "%s%s\"%s\"", (k > 0 ? valid_keys : ""),
+                    (k > 0 ? ", " : ""), key_names[k]);
+                free(valid_keys);
+                valid_keys = tmp;
```

Comment:
You could use `opal_argv_join()` to create this string.

Reply:
Didn't know this one either!, I think it's not suitable since key_names is not NULL-terminated, but the `opal_argv_join_range` is usable. I updated the code to use it, in the help text and in the descriptiton of the MCA parameter -- we lose a bit formatting e.g. the space after the comma, but I'd say it's okay.